### PR TITLE
chore: convert coffee-lex to TypeScript

### PR DESCRIPTION
**Pull request by eventualbuddha**
Per decaffeinate/decaffeinate#573. For now there should not be any functional or API changes.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -6,9 +6,9 @@ cache:
 notifications:
   email: false
 node_js:
+  - '7'
   - '6'
   - '4'
-  - '0.12'
```
Comment (eventualbuddha) [inline .travis.yml]: mocha-typescript uses `let` and possibly other es2015 features not present in `0.12`, so I figured it was time to let support for that version go.

File: package.json [hunk 2]
```diff
@@ -32,21 +32,16 @@
...
   "devDependencies": {
-    "babel-plugin-external-helpers": "^6.8.0",
-    "babel-plugin-syntax-class-properties": "^6.5.0",
-    "babel-plugin-syntax-flow": "^6.5.0",
-    "babel-plugin-transform-class-properties": "^6.19.0",
-    "babel-plugin-transform-flow-strip-types": "^6.7.0",
-    "babel-preset-es2015": "^6.18.0",
-    "babel-register": "^6.7.2",
-    "babelrc-rollup": "^3.0.0",
+    "@types/node": "^6.0.51",
```
Comment (eventualbuddha) [inline package.json]: We use `assert` and some other node libs. Nice to have these type checked!

File: src/SourceTokenList.ts [hunk 3]
```diff
@@ -37,7 +35,7 @@ export default class SourceTokenList {
    * Map each token to an element of an array.
    */
   map<T>(mapper: (token: SourceToken, index: SourceTokenListIndex, list: SourceTokenList) => T): Array<T> {
-    let result = [];
+    let result: Array<T> = [];
```
Comment (eventualbuddha) [inline src/SourceTokenList.ts]: This was the most common sort of change. TypeScript doesn't infer the type of the array based on usage, so you have to tell it.

File: src/SourceTokenList.ts [hunk 4]
```diff
@@ -271,18 +269,22 @@ export default class SourceTokenList {
...
       } else {
         let result = { done: false, value: this.tokenAtIndex(index) };
-        index = index.next();
+        let nextIndex = index.next();
+
+        if (!nextIndex) {
+          throw new Error(`unexpected null index before the end index`);
+        }
+
+        index = nextIndex;
```
Comment (eventualbuddha) [inline src/SourceTokenList.ts]: This is an example of TypeScript being stricter than Flow. Here, TypeScript sees that `tokenAtIndex` doesn't allow `null` but `index` could be set to `null` in `index = index.next();`.

**Comment by alangpierce** [comment]
Cool, thanks for annotating the interesting stuff. I just tried it out and I'm really impressed with the WebStorm support. Hopefully Flow support will get that good eventually.

**Comment by eventualbuddha** [comment]
Ah, yeah I'll do that for the other projects.

**Comment by alangpierce** [comment]
I have some thoughts on how to maybe set things up differently, but I'll move discussion to the original decaffeinate issue ( https://github.com/decaffeinate/decaffeinate/issues/573 ).

**Comment by eventualbuddha** [comment]
Your feedback is definitely welcome! I cobbled together what I did in coffee-lex based on what seemed to work, not with a deep understanding of the options available 😊

**Pull request merged**